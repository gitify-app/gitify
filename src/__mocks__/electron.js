window.Notification = function (title) {
  this.title = title;

  return {
    onclick: jest.fn(),
  };
};

window.Audio = class Audio {
  constructor(path) {
    this.path = path;
  }

  play() {}
};

window.localStorage = {
  store: {},
  getItem: function (key) {
    return this.store[key];
  },
  setItem: function (key, item) {
    this.store[key] = item;
  },
  removeItem: jest.fn(),
};

window.alert = jest.fn();

module.exports = {
  ipcRenderer: {
    send: jest.fn(),
    on: jest.fn(),
    sendSync: jest.fn(),
    invoke: jest.fn((channel, ...args) => {
      switch (channel) {
        case 'get-platform':
          return Promise.resolve('darwin');
        case 'gitify:version':
          return Promise.resolve('0.0.1');
        default:
          return Promise.reject(new Error(`Unknown channel: ${channel}`));
      }
    }),
  },
  shell: {
    openExternal: jest.fn(),
  },
};
