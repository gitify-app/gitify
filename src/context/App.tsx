import {
  type ReactNode,
  createContext,
  useCallback,
  useEffect,
  useMemo,
  useState,
} from 'react';

import { useInterval } from '../hooks/useInterval';
import { useNotifications } from '../hooks/useNotifications';
import {
  type AccountNotifications,
  type AuthOptions,
  type AuthState,
  type AuthTokenOptions,
  type GitifyError,
  type SettingsState,
  type Status,
  Theme,
} from '../types';
import { headNotifications } from '../utils/api/client';
import { addAccount, authGitHub, getToken, getUserData } from '../utils/auth';
import { setAutoLaunch, updateTrayTitle } from '../utils/comms';
import Constants from '../utils/constants';
import { getNotificationCount } from '../utils/notifications';
import { clearState, loadState, saveState } from '../utils/storage';
import { setTheme } from '../utils/theme';

const defaultAccounts: AuthState = {
  token: null,
  enterpriseAccounts: [],
  user: null,
};

export const defaultSettings: SettingsState = {
  participating: false,
  playSound: true,
  showNotifications: true,
  showBots: true,
  showNotificationsCountInTray: false,
  openAtStartup: false,
  theme: Theme.SYSTEM,
  detailedNotifications: false,
  markAsDoneOnOpen: false,
  showAccountHostname: false,
};

interface AppContextState {
  accounts: AuthState;
  isLoggedIn: boolean;
  login: () => void;
  loginEnterprise: (data: AuthOptions) => void;
  validateToken: (data: AuthTokenOptions) => void;
  logout: () => void;

  notifications: AccountNotifications[];
  status: Status;
  errorDetails: GitifyError;
  removeNotificationFromState: (id: string, hostname: string) => void;
  fetchNotifications: () => Promise<void>;
  markNotificationRead: (id: string, hostname: string) => Promise<void>;
  markNotificationDone: (id: string, hostname: string) => Promise<void>;
  unsubscribeNotification: (id: string, hostname: string) => Promise<void>;
  markRepoNotifications: (id: string, hostname: string) => Promise<void>;
  markRepoNotificationsDone: (id: string, hostname: string) => Promise<void>;

  settings: SettingsState;
  updateSetting: (
    name: keyof SettingsState,
    value: boolean | Theme | string | null,
  ) => void;
}

export const AppContext = createContext<Partial<AppContextState>>({});

export const AppProvider = ({ children }: { children: ReactNode }) => {
  const [accounts, setAccounts] = useState<AuthState>(defaultAccounts);
  const [settings, setSettings] = useState<SettingsState>(defaultSettings);
  const {
    fetchNotifications,
    notifications,
    errorDetails,
    status,
    removeNotificationFromState,
    markNotificationRead,
    markNotificationDone,
    unsubscribeNotification,
    markRepoNotifications,
    markRepoNotificationsDone,
  } = useNotifications();

  useEffect(() => {
    restoreSettings();
  }, []);

  useEffect(() => {
    setTheme(settings.theme as Theme);
  }, [settings.theme]);

  // biome-ignore lint/correctness/useExhaustiveDependencies: We only want fetchNotifications to be called for certain account or setting changes.
  useEffect(() => {
    fetchNotifications(accounts, settings);
  }, [
    settings.participating,
    settings.showBots,
    settings.detailedNotifications,
    accounts.token,
    accounts.enterpriseAccounts.length,
  ]);

  useInterval(() => {
    fetchNotifications(accounts, settings);
  }, Constants.FETCH_INTERVAL);

  // biome-ignore lint/correctness/useExhaustiveDependencies: We need to update tray title when settings or notifications changes.
  useEffect(() => {
    const count = getNotificationCount(notifications);

    if (settings.showNotificationsCountInTray && count > 0) {
      updateTrayTitle(count.toString());
    } else {
      updateTrayTitle();
    }
  }, [settings.showNotificationsCountInTray, notifications]);

  const updateSetting = useCallback(
    (name: keyof SettingsState, value: boolean | Theme) => {
      if (name === 'openAtStartup') {
        setAutoLaunch(value as boolean);
      }

      const newSettings = { ...settings, [name]: value };
      setSettings(newSettings);
      saveState(accounts, newSettings);
    },
    [accounts, settings],
  );

  const isLoggedIn = useMemo(() => {
    return !!accounts.token || accounts.enterpriseAccounts.length > 0;
  }, [accounts]);

  const login = useCallback(async () => {
    const { authCode } = await authGitHub();
    const { token } = await getToken(authCode);
    const hostname = Constants.DEFAULT_AUTH_OPTIONS.hostname;
    const user = await getUserData(token, hostname);
    const updatedAccounts = addAccount(accounts, token, hostname, user);
    setAccounts(updatedAccounts);
    saveState(updatedAccounts, settings);
  }, [accounts, settings]);

  const loginEnterprise = useCallback(
    async (data: AuthOptions) => {
      const { authOptions, authCode } = await authGitHub(data);
      const { token, hostname } = await getToken(authCode, authOptions);
      const updatedAccounts = addAccount(accounts, token, hostname);
      setAccounts(updatedAccounts);
      saveState(updatedAccounts, settings);
    },
    [accounts, settings],
  );

  const validateToken = useCallback(
    async ({ token, hostname }: AuthTokenOptions) => {
      await headNotifications(hostname, token);

      const user = await getUserData(token, hostname);
      const updatedAccounts = addAccount(accounts, token, hostname, user);
      setAccounts(updatedAccounts);
      saveState(updatedAccounts, settings);
    },
    [accounts, settings],
  );

  const logout = useCallback(() => {
    setAccounts(defaultAccounts);
    clearState();
  }, []);

  const restoreSettings = useCallback(() => {
    const existing = loadState();

    if (existing.accounts) {
      setAccounts({ ...defaultAccounts, ...existing.accounts });
    }

    if (existing.settings) {
      setSettings({ ...defaultSettings, ...existing.settings });
      return existing.settings;
    }
  }, []);

  const fetchNotificationsWithAccounts = useCallback(
    async () => await fetchNotifications(accounts, settings),
    [accounts, settings, notifications],
  );

  const markNotificationReadWithAccounts = useCallback(
    async (id: string, hostname: string) =>
      await markNotificationRead(accounts, id, hostname),
    [accounts, notifications],
  );

  const markNotificationDoneWithAccounts = useCallback(
    async (id: string, hostname: string) =>
      await markNotificationDone(accounts, id, hostname),
    [accounts, notifications],
  );

  const unsubscribeNotificationWithAccounts = useCallback(
    async (id: string, hostname: string) =>
      await unsubscribeNotification(accounts, id, hostname),
    [accounts, notifications],
  );

  const markRepoNotificationsWithAccounts = useCallback(
    async (repoSlug: string, hostname: string) =>
      await markRepoNotifications(accounts, repoSlug, hostname),
    [accounts, notifications],
  );

  const markRepoNotificationsDoneWithAccounts = useCallback(
    async (repoSlug: string, hostname: string) =>
      await markRepoNotificationsDone(accounts, repoSlug, hostname),
    [accounts, notifications],
  );

  return (
    <AppContext.Provider
      value={{
        accounts,
        isLoggedIn,
        login,
        loginEnterprise,
        validateToken,
        logout,

        notifications,
        status,
        errorDetails,
        removeNotificationFromState,
        fetchNotifications: fetchNotificationsWithAccounts,
        markNotificationRead: markNotificationReadWithAccounts,
        markNotificationDone: markNotificationDoneWithAccounts,
        unsubscribeNotification: unsubscribeNotificationWithAccounts,
        markRepoNotifications: markRepoNotificationsWithAccounts,
        markRepoNotificationsDone: markRepoNotificationsDoneWithAccounts,

        settings,
        updateSetting,
      }}
    >
      {children}
    </AppContext.Provider>
  );
};
