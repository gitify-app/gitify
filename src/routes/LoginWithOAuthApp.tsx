import { BookIcon, PersonIcon, SignInIcon } from '@primer/octicons-react';
import { type FC, useCallback, useContext } from 'react';
import { Form, type FormRenderProps } from 'react-final-form';
import { Header } from '../components/Header';
import { Button } from '../components/buttons/Button';
import { FieldInput } from '../components/fields/FieldInput';
import { AppContext } from '../context/App';
import {
  type ClientID,
  type ClientSecret,
  type Hostname,
  Size,
  type Token,
} from '../types';
import type { LoginOAuthAppOptions } from '../utils/auth/types';
import {
  getNewOAuthAppURL,
  isValidClientId,
  isValidHostname,
  isValidToken,
} from '../utils/auth/utils';
import Constants from '../utils/constants';

interface IValues {
  hostname?: Hostname;
  clientId?: ClientID;
  clientSecret?: ClientSecret;
}

interface IFormErrors {
  hostname?: string;
  clientId?: string;
  clientSecret?: string;
}

export const validate = (values: IValues): IFormErrors => {
  const errors: IFormErrors = {};

  if (!values.hostname) {
    errors.hostname = 'Required';
  } else if (!isValidHostname(values.hostname)) {
    errors.hostname = 'Invalid hostname.';
  }

  if (!values.clientId) {
    errors.clientId = 'Required';
  } else if (!isValidClientId(values.clientId)) {
    errors.clientId = 'Invalid client id.';
  }

  if (!values.clientSecret) {
    errors.clientSecret = 'Required';
  } else if (!isValidToken(values.clientSecret as unknown as Token)) {
    errors.clientSecret = 'Invalid client secret.';
  }

  return errors;
};

export const LoginWithOAuthApp: FC = () => {
  const { loginWithOAuthApp } = useContext(AppContext);

  const renderForm = (formProps: FormRenderProps) => {
    const { handleSubmit, submitting, pristine, values } = formProps;

    return (
      <form onSubmit={handleSubmit}>
        <FieldInput
          name="hostname"
          label="Hostname"
          placeholder="github.company.com"
          helpText={
            <div className="mb-1">
              <Button
                label="Create new OAuth App"
                disabled={!values.hostname}
                icon={{ icon: PersonIcon, size: Size.XSMALL }}
                url={getNewOAuthAppURL(values.hostname)}
              >
                Create new OAuth App
              </Button>
              <span className="mx-1">on GitHub then paste your</span>
              <span className="italic">client id and client secret</span> below.
            </div>
          }
        />

        <FieldInput name="clientId" label="Client ID" placeholder="123456789" />

        <FieldInput
          name="clientSecret"
          label="Client Secret"
          placeholder="ABC123DEF456"
        />

        <div className="flex items-end justify-between">
          <Button
            name="Docs"
            label="GitHub Docs"
            className="mt-2"
            icon={{ icon: BookIcon, size: Size.XSMALL }}
            url={Constants.GITHUB_DOCS.OAUTH_URL}
          />

          <Button
            name="Login"
            label="Login"
            className="mt-2 px-4 py-2 !text-sm"
            icon={{ icon: SignInIcon, size: Size.MEDIUM }}
            disabled={submitting || pristine}
            type="submit"
          />
        </div>
      </form>
    );
  };

  const login = useCallback(
    async (data: IValues) => {
      try {
        await loginWithOAuthApp(data as LoginOAuthAppOptions);
      } catch (err) {
        // Skip
      }
    },
    [loginWithOAuthApp],
  );

  return (
    <div>
      <Header icon={PersonIcon}>Login with OAuth App</Header>
      <div className="px-8">
        <Form
          initialValues={{
            hostname: '',
            clientId: '',
            clientSecret: '',
          }}
          onSubmit={login}
          validate={validate}
        >
          {renderForm}
        </Form>
      </div>
    </div>
  );
};
