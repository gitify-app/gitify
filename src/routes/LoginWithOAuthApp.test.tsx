import { fireEvent, render, screen } from '@testing-library/react';
import { shell } from 'electron';
import { MemoryRouter } from 'react-router-dom';
import { AppContext } from '../context/App';
import type { AuthState, ClientID, ClientSecret, Hostname } from '../types';
import { LoginWithOAuthApp, validate } from './LoginWithOAuthApp';

const mockNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate,
}));

describe('routes/LoginWithOAuthApp.tsx', () => {
  const openExternalMock = jest.spyOn(shell, 'openExternal');

  const mockAuth: AuthState = {
    accounts: [],
  };

  beforeEach(() => {
    openExternalMock.mockReset();
    mockNavigate.mockReset();
  });

  it('renders correctly', () => {
    const tree = render(
      <AppContext.Provider value={{ auth: mockAuth }}>
        <MemoryRouter>
          <LoginWithOAuthApp />
        </MemoryRouter>
      </AppContext.Provider>,
    );

    expect(tree).toMatchSnapshot();
  });

  it('let us go back', () => {
    render(
      <AppContext.Provider value={{ auth: mockAuth }}>
        <MemoryRouter>
          <LoginWithOAuthApp />
        </MemoryRouter>
      </AppContext.Provider>,
    );

    fireEvent.click(screen.getByLabelText('Go Back'));
    expect(mockNavigate).toHaveBeenNthCalledWith(1, -1);
  });

  it('should validate the form values', () => {
    const emptyValues = {
      hostname: null,
      clientId: null,
      clientSecret: null,
    };

    let values = {
      ...emptyValues,
    };
    expect(validate(values).hostname).toBe('Required');
    expect(validate(values).clientId).toBe('Required');
    expect(validate(values).clientSecret).toBe('Required');

    values = {
      ...emptyValues,
      hostname: 'hello' as Hostname,
      clientId: '!@£INVALID-.1' as ClientID,
      clientSecret: '!@£INVALID-.1' as ClientSecret,
    };
    expect(validate(values).hostname).toBe('Invalid hostname.');
    expect(validate(values).clientId).toBe('Invalid client id.');
    expect(validate(values).clientSecret).toBe('Invalid client secret.');
  });

  describe("'Create new OAuth App' button", () => {
    it('should be disabled if no hostname configured', async () => {
      render(
        <AppContext.Provider value={{ auth: mockAuth }}>
          <MemoryRouter>
            <LoginWithOAuthApp />
          </MemoryRouter>
        </AppContext.Provider>,
      );

      fireEvent.click(screen.getByText('Create new OAuth App'));

      expect(openExternalMock).toHaveBeenCalledTimes(0);
    });

    it('should open in browser if hostname configured', async () => {
      render(
        <AppContext.Provider value={{ auth: mockAuth }}>
          <MemoryRouter>
            <LoginWithOAuthApp />
          </MemoryRouter>
        </AppContext.Provider>,
      );

      fireEvent.change(screen.getByLabelText('Hostname'), {
        target: { value: 'company.github.com' },
      });

      fireEvent.click(screen.getByText('Create new OAuth App'));

      expect(openExternalMock).toHaveBeenCalledTimes(1);
    });
  });

  it('should render the form with errors', () => {
    render(
      <AppContext.Provider value={{ auth: mockAuth }}>
        <MemoryRouter>
          <LoginWithOAuthApp />
        </MemoryRouter>
      </AppContext.Provider>,
    );

    fireEvent.change(screen.getByLabelText('Hostname'), {
      target: { value: 'test' },
    });
    fireEvent.change(screen.getByLabelText('Client ID'), {
      target: { value: '123' },
    });
    fireEvent.change(screen.getByLabelText('Client Secret'), {
      target: { value: 'abc' },
    });

    fireEvent.submit(screen.getByTitle('Login'));

    expect(screen.getByText('Invalid hostname.')).toBeTruthy();
    expect(screen.getByText('Invalid client id.')).toBeTruthy();
    expect(screen.getByText('Invalid client secret.')).toBeTruthy();
  });

  it('should open help docs in the browser', async () => {
    render(
      <AppContext.Provider value={{ auth: mockAuth }}>
        <MemoryRouter>
          <LoginWithOAuthApp />
        </MemoryRouter>
      </AppContext.Provider>,
    );

    fireEvent.click(screen.getByLabelText('GitHub Docs'));

    expect(openExternalMock).toHaveBeenCalledTimes(1);
  });
});
