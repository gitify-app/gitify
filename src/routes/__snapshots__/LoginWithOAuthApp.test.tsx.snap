// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`routes/LoginWithOAuthApp.tsx renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="flex-1 bg-white dark:bg-gray-dark dark:text-white"
      >
        <div
          class="mx-8 mt-4 flex items-center justify-between py-2"
        >
          <button
            class="focus:outline-none"
            title="Go Back"
            type="button"
          >
            <svg
              aria-label="Go Back"
              class="hover:text-gray-400"
              fill="currentColor"
              focusable="false"
              height="20"
              role="img"
              style="display: inline-block; user-select: none; vertical-align: text-bottom; overflow: visible;"
              viewBox="0 0 16 16"
              width="20"
            >
              <path
                d="M7.78 12.53a.75.75 0 0 1-1.06 0L2.47 8.28a.75.75 0 0 1 0-1.06l4.25-4.25a.751.751 0 0 1 1.042.018.751.751 0 0 1 .018 1.042L4.81 7h7.44a.75.75 0 0 1 0 1.5H4.81l2.97 2.97a.75.75 0 0 1 0 1.06Z"
              />
            </svg>
          </button>
          <h3
            class="justify-center text-lg font-semibold"
          >
            <svg
              aria-hidden="true"
              class="mr-2"
              fill="currentColor"
              focusable="false"
              height="20"
              style="display: inline-block; user-select: none; vertical-align: text-bottom; overflow: visible;"
              viewBox="0 0 16 16"
              width="20"
            >
              <path
                d="M10.561 8.073a6.005 6.005 0 0 1 3.432 5.142.75.75 0 1 1-1.498.07 4.5 4.5 0 0 0-8.99 0 .75.75 0 0 1-1.498-.07 6.004 6.004 0 0 1 3.431-5.142 3.999 3.999 0 1 1 5.123 0ZM10.5 5a2.5 2.5 0 1 0-5 0 2.5 2.5 0 0 0 5 0Z"
              />
            </svg>
            Login with OAuth App
          </h3>
        </div>
        <div
          class="flex-1 px-8"
        >
          <form>
            <div
              class="mb-4"
            >
              <label
                class="mb-2 block text-sm font-semibold tracking-wide text-gray-dark"
                for="hostname"
              >
                Hostname
              </label>
              <input
                class="mb-2 block w-full appearance-none rounded border border-red-500 bg-gray-100 px-4 py-1.5 text-sm focus:bg-gray-200 focus:outline-none dark:text-gray-800"
                id="hostname"
                name="hostname"
                placeholder="github.company.com"
                type="text"
                value=""
              />
              <div
                class="mt-3 text-xs text-gray-700 dark:text-gray-200"
              >
                <div
                  class="mb-1"
                >
                  <button
                    aria-label="Create new OAuth App"
                    class="rounded bg-gray-300 font-semibold text-xs text-center hover:bg-gray-500 hover:text-white dark:text-black focus:outline-none cursor-pointer px-2 py-1"
                    disabled=""
                    title="Create new OAuth App"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="mr-1"
                      fill="currentColor"
                      focusable="false"
                      height="12"
                      style="display: inline-block; user-select: none; vertical-align: text-bottom; overflow: visible;"
                      viewBox="0 0 16 16"
                      width="12"
                    >
                      <path
                        d="M10.561 8.073a6.005 6.005 0 0 1 3.432 5.142.75.75 0 1 1-1.498.07 4.5 4.5 0 0 0-8.99 0 .75.75 0 0 1-1.498-.07 6.004 6.004 0 0 1 3.431-5.142 3.999 3.999 0 1 1 5.123 0ZM10.5 5a2.5 2.5 0 1 0-5 0 2.5 2.5 0 0 0 5 0Z"
                      />
                    </svg>
                    Create new OAuth App
                  </button>
                  <span
                    class="mx-1"
                  >
                    on GitHub then paste your
                  </span>
                  <span
                    class="italic"
                  >
                    client id and client secret
                  </span>
                   below.
                </div>
              </div>
            </div>
            <div
              class="mb-4"
            >
              <label
                class="mb-2 block text-sm font-semibold tracking-wide text-gray-dark"
                for="clientId"
              >
                Client ID
              </label>
              <input
                class="mb-2 block w-full appearance-none rounded border border-red-500 bg-gray-100 px-4 py-1.5 text-sm focus:bg-gray-200 focus:outline-none dark:text-gray-800"
                id="clientId"
                name="clientId"
                placeholder="123456789"
                type="text"
                value=""
              />
            </div>
            <div
              class="mb-4"
            >
              <label
                class="mb-2 block text-sm font-semibold tracking-wide text-gray-dark"
                for="clientSecret"
              >
                Client Secret
              </label>
              <input
                class="mb-2 block w-full appearance-none rounded border border-red-500 bg-gray-100 px-4 py-1.5 text-sm focus:bg-gray-200 focus:outline-none dark:text-gray-800"
                id="clientSecret"
                name="clientSecret"
                placeholder="ABC123DEF456"
                type="text"
                value=""
              />
            </div>
            <div
              class="flex items-end justify-between"
            >
              <button
                aria-label="GitHub Docs"
                class="mt-2 rounded bg-gray-300 font-semibold text-xs text-center hover:bg-gray-500 hover:text-white dark:text-black focus:outline-none cursor-pointer px-2 py-1"
                title="GitHub Docs"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="mr-1"
                  fill="currentColor"
                  focusable="false"
                  height="12"
                  style="display: inline-block; user-select: none; vertical-align: text-bottom; overflow: visible;"
                  viewBox="0 0 16 16"
                  width="12"
                >
                  <path
                    d="M0 1.75A.75.75 0 0 1 .75 1h4.253c1.227 0 2.317.59 3 1.501A3.743 3.743 0 0 1 11.006 1h4.245a.75.75 0 0 1 .75.75v10.5a.75.75 0 0 1-.75.75h-4.507a2.25 2.25 0 0 0-1.591.659l-.622.621a.75.75 0 0 1-1.06 0l-.622-.621A2.25 2.25 0 0 0 5.258 13H.75a.75.75 0 0 1-.75-.75Zm7.251 10.324.004-5.073-.002-2.253A2.25 2.25 0 0 0 5.003 2.5H1.5v9h3.757a3.75 3.75 0 0 1 1.994.574ZM8.755 4.75l-.004 7.322a3.752 3.752 0 0 1 1.992-.572H14.5v-9h-3.495a2.25 2.25 0 0 0-2.25 2.25Z"
                  />
                </svg>
                Docs
              </button>
              <button
                aria-label="Login"
                class="mt-2 px-4 py-2 !text-sm rounded bg-gray-300 font-semibold text-xs text-center hover:bg-gray-500 hover:text-white dark:text-black focus:outline-none cursor-pointer px-2 py-1"
                disabled=""
                title="Login"
                type="submit"
              >
                <svg
                  aria-hidden="true"
                  class="mr-1"
                  fill="currentColor"
                  focusable="false"
                  height="16"
                  style="display: inline-block; user-select: none; vertical-align: text-bottom; overflow: visible;"
                  viewBox="0 0 16 16"
                  width="16"
                >
                  <path
                    d="M2 2.75C2 1.784 2.784 1 3.75 1h2.5a.75.75 0 0 1 0 1.5h-2.5a.25.25 0 0 0-.25.25v10.5c0 .138.112.25.25.25h2.5a.75.75 0 0 1 0 1.5h-2.5A1.75 1.75 0 0 1 2 13.25Zm6.56 4.5h5.69a.75.75 0 0 1 0 1.5H8.56l1.97 1.97a.749.749 0 0 1-.326 1.275.749.749 0 0 1-.734-.215L6.22 8.53a.75.75 0 0 1 0-1.06l3.25-3.25a.749.749 0 0 1 1.275.326.749.749 0 0 1-.215.734Z"
                  />
                </svg>
                Login
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex-1 bg-white dark:bg-gray-dark dark:text-white"
    >
      <div
        class="mx-8 mt-4 flex items-center justify-between py-2"
      >
        <button
          class="focus:outline-none"
          title="Go Back"
          type="button"
        >
          <svg
            aria-label="Go Back"
            class="hover:text-gray-400"
            fill="currentColor"
            focusable="false"
            height="20"
            role="img"
            style="display: inline-block; user-select: none; vertical-align: text-bottom; overflow: visible;"
            viewBox="0 0 16 16"
            width="20"
          >
            <path
              d="M7.78 12.53a.75.75 0 0 1-1.06 0L2.47 8.28a.75.75 0 0 1 0-1.06l4.25-4.25a.751.751 0 0 1 1.042.018.751.751 0 0 1 .018 1.042L4.81 7h7.44a.75.75 0 0 1 0 1.5H4.81l2.97 2.97a.75.75 0 0 1 0 1.06Z"
            />
          </svg>
        </button>
        <h3
          class="justify-center text-lg font-semibold"
        >
          <svg
            aria-hidden="true"
            class="mr-2"
            fill="currentColor"
            focusable="false"
            height="20"
            style="display: inline-block; user-select: none; vertical-align: text-bottom; overflow: visible;"
            viewBox="0 0 16 16"
            width="20"
          >
            <path
              d="M10.561 8.073a6.005 6.005 0 0 1 3.432 5.142.75.75 0 1 1-1.498.07 4.5 4.5 0 0 0-8.99 0 .75.75 0 0 1-1.498-.07 6.004 6.004 0 0 1 3.431-5.142 3.999 3.999 0 1 1 5.123 0ZM10.5 5a2.5 2.5 0 1 0-5 0 2.5 2.5 0 0 0 5 0Z"
            />
          </svg>
          Login with OAuth App
        </h3>
      </div>
      <div
        class="flex-1 px-8"
      >
        <form>
          <div
            class="mb-4"
          >
            <label
              class="mb-2 block text-sm font-semibold tracking-wide text-gray-dark"
              for="hostname"
            >
              Hostname
            </label>
            <input
              class="mb-2 block w-full appearance-none rounded border border-red-500 bg-gray-100 px-4 py-1.5 text-sm focus:bg-gray-200 focus:outline-none dark:text-gray-800"
              id="hostname"
              name="hostname"
              placeholder="github.company.com"
              type="text"
              value=""
            />
            <div
              class="mt-3 text-xs text-gray-700 dark:text-gray-200"
            >
              <div
                class="mb-1"
              >
                <button
                  aria-label="Create new OAuth App"
                  class="rounded bg-gray-300 font-semibold text-xs text-center hover:bg-gray-500 hover:text-white dark:text-black focus:outline-none cursor-pointer px-2 py-1"
                  disabled=""
                  title="Create new OAuth App"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="mr-1"
                    fill="currentColor"
                    focusable="false"
                    height="12"
                    style="display: inline-block; user-select: none; vertical-align: text-bottom; overflow: visible;"
                    viewBox="0 0 16 16"
                    width="12"
                  >
                    <path
                      d="M10.561 8.073a6.005 6.005 0 0 1 3.432 5.142.75.75 0 1 1-1.498.07 4.5 4.5 0 0 0-8.99 0 .75.75 0 0 1-1.498-.07 6.004 6.004 0 0 1 3.431-5.142 3.999 3.999 0 1 1 5.123 0ZM10.5 5a2.5 2.5 0 1 0-5 0 2.5 2.5 0 0 0 5 0Z"
                    />
                  </svg>
                  Create new OAuth App
                </button>
                <span
                  class="mx-1"
                >
                  on GitHub then paste your
                </span>
                <span
                  class="italic"
                >
                  client id and client secret
                </span>
                 below.
              </div>
            </div>
          </div>
          <div
            class="mb-4"
          >
            <label
              class="mb-2 block text-sm font-semibold tracking-wide text-gray-dark"
              for="clientId"
            >
              Client ID
            </label>
            <input
              class="mb-2 block w-full appearance-none rounded border border-red-500 bg-gray-100 px-4 py-1.5 text-sm focus:bg-gray-200 focus:outline-none dark:text-gray-800"
              id="clientId"
              name="clientId"
              placeholder="123456789"
              type="text"
              value=""
            />
          </div>
          <div
            class="mb-4"
          >
            <label
              class="mb-2 block text-sm font-semibold tracking-wide text-gray-dark"
              for="clientSecret"
            >
              Client Secret
            </label>
            <input
              class="mb-2 block w-full appearance-none rounded border border-red-500 bg-gray-100 px-4 py-1.5 text-sm focus:bg-gray-200 focus:outline-none dark:text-gray-800"
              id="clientSecret"
              name="clientSecret"
              placeholder="ABC123DEF456"
              type="text"
              value=""
            />
          </div>
          <div
            class="flex items-end justify-between"
          >
            <button
              aria-label="GitHub Docs"
              class="mt-2 rounded bg-gray-300 font-semibold text-xs text-center hover:bg-gray-500 hover:text-white dark:text-black focus:outline-none cursor-pointer px-2 py-1"
              title="GitHub Docs"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="mr-1"
                fill="currentColor"
                focusable="false"
                height="12"
                style="display: inline-block; user-select: none; vertical-align: text-bottom; overflow: visible;"
                viewBox="0 0 16 16"
                width="12"
              >
                <path
                  d="M0 1.75A.75.75 0 0 1 .75 1h4.253c1.227 0 2.317.59 3 1.501A3.743 3.743 0 0 1 11.006 1h4.245a.75.75 0 0 1 .75.75v10.5a.75.75 0 0 1-.75.75h-4.507a2.25 2.25 0 0 0-1.591.659l-.622.621a.75.75 0 0 1-1.06 0l-.622-.621A2.25 2.25 0 0 0 5.258 13H.75a.75.75 0 0 1-.75-.75Zm7.251 10.324.004-5.073-.002-2.253A2.25 2.25 0 0 0 5.003 2.5H1.5v9h3.757a3.75 3.75 0 0 1 1.994.574ZM8.755 4.75l-.004 7.322a3.752 3.752 0 0 1 1.992-.572H14.5v-9h-3.495a2.25 2.25 0 0 0-2.25 2.25Z"
                />
              </svg>
              Docs
            </button>
            <button
              aria-label="Login"
              class="mt-2 px-4 py-2 !text-sm rounded bg-gray-300 font-semibold text-xs text-center hover:bg-gray-500 hover:text-white dark:text-black focus:outline-none cursor-pointer px-2 py-1"
              disabled=""
              title="Login"
              type="submit"
            >
              <svg
                aria-hidden="true"
                class="mr-1"
                fill="currentColor"
                focusable="false"
                height="16"
                style="display: inline-block; user-select: none; vertical-align: text-bottom; overflow: visible;"
                viewBox="0 0 16 16"
                width="16"
              >
                <path
                  d="M2 2.75C2 1.784 2.784 1 3.75 1h2.5a.75.75 0 0 1 0 1.5h-2.5a.25.25 0 0 0-.25.25v10.5c0 .138.112.25.25.25h2.5a.75.75 0 0 1 0 1.5h-2.5A1.75 1.75 0 0 1 2 13.25Zm6.56 4.5h5.69a.75.75 0 0 1 0 1.5H8.56l1.97 1.97a.749.749 0 0 1-.326 1.275.749.749 0 0 1-.734-.215L6.22 8.53a.75.75 0 0 1 0-1.06l3.25-3.25a.749.749 0 0 1 1.275.326.749.749 0 0 1-.215.734Z"
                />
              </svg>
              Login
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
